File: backend/src/main/java/com/example/demo/controller/NotificationController.java
package com.example.demo.controller;

import com.example.demo.repository.SseEmitters;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

@Slf4j
@RestController
@RequestMapping("/api/notifications")
@RequiredArgsConstructor
public class NotificationController {

    private final SseEmitters sseEmitters;

    @GetMapping(value = "/subscribe", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public SseEmitter subscribe() {
        // Timeout: 60 seconds (can be adjusted)
        SseEmitter emitter = new SseEmitter(60 * 1000L);
        return sseEmitters.add(emitter);
    }
}

File: backend/src/main/kotlin/com/example/demo/controller/NotificationController.kt
package com.example.demo.controller

import com.example.demo.dto.NotificationMessage
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class NotificationController(private val template: SimpMessagingTemplate) {

    @PostMapping("/api/notify")
    fun sendNotification(@RequestBody message: NotificationMessage) {
        template.convertAndSend("/topic/notifications", message)
    }
}
